File: android/DFM/src/main/java/com/dontflymoney/baseactivity/Navigation.java
package com.dontflymoney.baseactivity;

import android.content.Intent;
import android.os.Bundle;

import com.dontflymoney.api.InternalRequest;
import com.dontflymoney.api.Step;
import com.dontflymoney.auth.Authentication;
import com.dontflymoney.view.LoginActivity;
import com.dontflymoney.view.SettingsActivity;

import org.json.JSONException;

public class Navigation
{
	private SmartActivity activity;
	private Authentication authentication;
	
	Navigation(SmartActivity activity, Authentication authentication)
	{
		this.activity = activity;
		this.authentication = authentication;
	}
	
	
	
	public void redirect(Class<?> activityClass)
	{
		Intent intent = new Intent(activity, activityClass);
		activity.startActivity(intent);
	}

	public void redirectWithExtras()
	{
		Bundle extras = activity.getIntent().getExtras();
		final Class<?> parent = (Class<?>)extras.get("__parent");
		extras.remove("__parent");

		Intent intent = new Intent(activity, parent);
		intent.putExtras(extras);

		activity.startActivity(intent);
	}






	void logout()
	{
		InternalRequest request = new InternalRequest(activity, "Users/Logout");
		request.AddParameter("ticket", authentication.Get());
		boolean tryResult = request.Post(Step.Logout);

		if (tryResult)
		{
			authentication.Clear();
			redirect(LoginActivity.class);
		}
	}
	
	void back()
	{
		activity.finish();
	}

	void close()
	{
		Intent intent = new Intent(activity, LoginActivity.class);
		intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
		intent.putExtra("EXIT", true);
		activity.startActivity(intent);
	}

	void goToSettings()
	{
		Intent intent = new Intent(activity, SettingsActivity.class);

		intent.putExtras(activity.getIntent());
		intent.putExtra("__parent", activity.getClass());

		activity.startActivity(intent);
	}
	
	
}

File: android/DFM/src/main/java/com/dontflymoney/baseactivity/Navigation.kt
package com.dontflymoney.baseactivity

import android.content.Intent
import android.os.Bundle
import com.dontflymoney.activityObjects.SmartStatic

import com.dontflymoney.api.InternalRequest
import com.dontflymoney.api.Step
import com.dontflymoney.userdata.Authentication
import com.dontflymoney.view.LoginActivity
import com.dontflymoney.view.SettingsActivity

class Navigation<T : SmartStatic> internal constructor(private val activity: SmartActivity<T>, private val authentication: Authentication) {


    fun redirect(activityClass: Class<*>) {
        val intent = Intent(activity, activityClass)
        activity.startActivity(intent)
    }

    fun redirectWithExtras() {
        val extras = activity.intent.extras
        val parent = extras.get("__parent") as Class<*>
        extras.remove("__parent")

        val intent = Intent(activity, parent)
        intent.putExtras(extras)

        activity.startActivity(intent)
    }


    internal fun logout() {
        val request = InternalRequest(activity, "Users/Logout")
        request.AddParameter("ticket", authentication.Get()!!)
        val tryResult = request.Post(Step.Logout)

        if (tryResult) {
            authentication.Clear()
            redirect(LoginActivity::class.java)
        }
    }

    internal fun back() {
        activity.finish()
    }

    internal fun close() {
        val intent = Intent(activity, LoginActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        intent.putExtra("EXIT", true)
        activity.startActivity(intent)
    }

    internal fun goToSettings() {
        val intent = Intent(activity, SettingsActivity::class.java)

        intent.putExtras(activity.intent)
        intent.putExtra("__parent", activity.javaClass)

        activity.startActivity(intent)
    }


}
